src/Adapter/Image/Uploader/ProductImageUploader.php

```diff
@@ -0,0 +1,140 @@
+
+use ErrorException;
+use Image;
+use ImageType;
+use PrestaShop\PrestaShop\Adapter\Image\ImageGenerator;
+use PrestaShop\PrestaShop\Adapter\Product\ProductImagePathFactory;
+use PrestaShop\PrestaShop\Core\Configuration\UploadSizeConfigurationInterface;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Image\Exception\CannotUnlinkImageException;
+
+/**
+ * Uploads product image to filesystem
+ */
+class ProductImageUploader extends AbstractImageUploader
+{
+    /**
+     * @var UploadSizeConfigurationInterface
+     */
+    private $uploadSizeConfiguration;
+
+    /**
+     * @var ProductImagePathFactory
+     */
+    private $productImagePathFactory;
+
+    /**
+     * @var int
+     */
+    private $contextShopId;
+
+    /**
+     * @var ImageGenerator
+     */
+    private $imageGenerator;
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @param UploadSizeConfigurationInterface $uploadSizeConfiguration
+     * @param ProductImagePathFactory $productImagePathFactory
+     * @param int $contextShopId
+     * @param ImageGenerator $imageGenerator
+     * @param HookDispatcherInterface $hookDispatcher
+     */
+    public function __construct(
+        UploadSizeConfigurationInterface $uploadSizeConfiguration,
+        ProductImagePathFactory $productImagePathFactory,
+        int $contextShopId,
+        ImageGenerator $imageGenerator,
+        HookDispatcherInterface $hookDispatcher
+    ) {
+        $this->uploadSizeConfiguration = $uploadSizeConfiguration;
+        $this->productImagePathFactory = $productImagePathFactory;
+        $this->contextShopId = $contextShopId;
+        $this->imageGenerator = $imageGenerator;
+        $this->hookDispatcher = $hookDispatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function upload(Image $image, string $filePath): void
+    {
+        $this->productImagePathFactory->createDestinationDirectory($image);
+        $destinationPath = $this->productImagePathFactory->getBasePath($image);
+        $this->uploadFromTemp($filePath, $destinationPath);
+        $this->imageGenerator->generateImagesByTypes($destinationPath, ImageType::getImagesTypes('products'));
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionWatermark',
+            ['id_image' => (int) $image->id, 'id_product' => (int) $image->id_product]
+        );
+
+        $this->deleteCachedImages($image);
+    }
+
+    /**
+     * @param Image $image
+     *
+     * @throws CannotUnlinkImageException
+     */
+    private function deleteCachedImages(Image $image): void
+    {
+        $productId = (int) $image->id_product;
+
+        $cachedImages = [
+            $this->productImagePathFactory->getCachedCover($productId, $this->contextShopId),
+            $this->productImagePathFactory->getCachedThumbnail($productId),
+        ];
+
+        foreach ($cachedImages as $cachedImage) {
```

Comment:
I'm wondering if we should use something similar to Symfony `Filesystem::box()` https://github.com/symfony/filesystem/blob/5.x/Filesystem.php#L716 like we discussed 🤔 

maybe in the future